internal/controller/accessrequest_controller.go

```diff
@@ -159,76 +154,246 @@ func buildResult(status api.Status, ar *api.AccessRequest) ctrl.Result {
+	details := ""
+	status, err := r.grantArgoCDAccess(ctx, ar)
 	if err != nil {
-		return "", fmt.Errorf("error updating AppProject RBAC: %w", err)
+		details = fmt.Sprintf("Error granting Argo CD Access: %s", err)
 	}
-	// only update the status if transitioning to granted
-	if ar.Status.RequestState != api.GrantedStatus {
-		err = r.updateStatusWithRetry(ctx, ar, api.GrantedStatus, resp.Message)
+	// only update status if the current state is different
+	if ar.Status.RequestState != status {
+		err = r.updateStatus(ctx, ar, status, details)
 		if err != nil {
 			return "", fmt.Errorf("error updating access request status to granted: %w", err)
 		}
 	}
-	return api.GrantedStatus, nil
+	return status, nil
 }
 
+// updateStatusWithRetry will retrieve the latest AccessRequest state before
+// attempting to update its status. In case of conflict error, it will retry
+// using the DefaultRetry backoff which has the following configs:
+//
+//	Steps: 5, Duration: 10 milliseconds, Factor: 1.0, Jitter: 0.1
 func (r *AccessRequestReconciler) updateStatusWithRetry(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
 		if err != nil {
 			return err
 		}
-		// if it is already updated skip
-		if ar.Status.RequestState == status {
-			return nil
+		return r.updateStatus(ctx, ar, status, details)
+	})
+}
+
+// updateStatus will update the given AccessRequest status field with the
+// given status and details.
+func (r *AccessRequestReconciler) updateStatus(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
+	// if it is already updated skip
+	if ar.Status.RequestState == status {
+		return nil
+	}
+	ar.UpdateStatus(status, details)
+	return r.Status().Update(ctx, ar)
+}
+
+// removeArgoCDAccess will remove the subjects in the given AccessRequest from
+// the given ar.TargetRoleName from the Argo CD project referenced in the
+// ar.Spec.AppProject. The AppProject update will be executed via a patch with
+// optimistic lock enabled. It will retry in case of AppProject conflict is
+// identied.
+func (r *AccessRequestReconciler) removeArgoCDAccess(ctx context.Context, ar *api.AccessRequest) error {
+	logger := log.FromContext(ctx)
+	logger.Info("Removing Argo CD Access")
+	project := &argocd.AppProject{}
+	objKey := client.ObjectKey{
+		Namespace: ar.Spec.AppProject.Namespace,
+		Name:      ar.Spec.AppProject.Name,
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, objKey, project)
+		if err != nil {
+			e := fmt.Errorf("error getting Argo CD Project %s/%s: %w", objKey.Namespace, objKey.Name, err)
+			return client.IgnoreNotFound(e)
+		}
+		patch := client.MergeFromWithOptions(project.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+		logger.Debug("Removing subjects from role")
+		removeSubjectsFromRole(project, ar)
```

Comment:
It should also remove/revoke any JWT tokens that were created. 
you can add a TODO in the code if you plan to implement it in another PR.

Reply:
added TODO